plaso/parsers/postgresql.py

```diff
@@ -0,0 +1,183 @@
+
+  _LOG_LEVEL = (
+    pyparsing.Word(string.ascii_uppercase) +
+    pyparsing.Suppress(':')
+  ).setResultsName('log_level')
+
+  _LOG_LINE_END = _BODY_END = pyparsing.StringEnd() | _DATE_TIME
+
+  _LOG_LINE = (
+    _DATE_TIME +
+    _PID +
+    pyparsing.Optional(_USER_AND_DATABASE) +
+    _LOG_LEVEL +
+    pyparsing.SkipTo(_LOG_LINE_END).setResultsName('log_line') +
+    pyparsing.ZeroOrMore(pyparsing.lineEnd())
+  )
+
+  LINE_STRUCTURES = [
+    ('logline', _LOG_LINE),
+  ]
+
+  def _ConvertTimeString(self, structure):
+    """Converts the structure to a datetime object.
+
+    The date and time values are formatted as:
+    "2022-04-12 00:16:05.526 UTC".
+
+    Args:
+      structure (pyparsing.ParseResults): structure of tokens derived from a
+          line of a text file, that contains the time elements.
+
+    Returns:
+      datetime: datetime object.
+
+    Raises:
+      ValueError: if the structure cannot be converted into a datetime.
+    """
+    try:
+      dts = '{0:d}-{1:d}-{2:d} {3:d}:{4:d}:{5:d}.{6:d} {7:s}'.format(
+        structure['year'], structure['month'], structure['day_of_month'],
+        structure['time']['hours'], structure['time']['minutes'],
+        structure['time']['seconds'], structure.get('microseconds', [0])[0],
+        structure['time_zone'])
+    except (TypeError, ValueError) as exception:
+      raise ValueError(
+          'unable to format date time string with error: {0!s}.'.format(
+              exception))
+    return parser.parse(timestr=dts)
+
+
+  def ParseRecord(self, parser_mediator, key, structure):
+    """Parses a record and produces a PostgreSQL event.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      key (str): name of the parsed structure.
+      structure (pyparsing.ParseResults): elements parsed from the file.
+
+    Raises:
+      ParseError: when the structure type is unknown.
+    """
+    if key != 'logline':
+      raise errors.ParseError(
+          'Unable to parse record, unknown structure: {0:s}'.format(key))
+
+    event_data = PostgreSQLEventData()
+    event_data.pid = ''.join(
+      [str(x) for x in self._GetValueFromStructure(structure, 'pid')])
+
+    log_level = self._GetValueFromStructure(structure, 'log_level')
+    if log_level and len(log_level) == 1:
+      event_data.log_level = log_level[0]
+    else:
+      parser_mediator.ProduceExtractionWarning('no log level found')
+      return
+
+    user_and_database = self._GetValueFromStructure(
+      structure, 'user_and_database')
+    if user_and_database:
+      event_data.user = ''.join(user_and_database)
+
+    log_line = self._GetValueFromStructure(structure, 'log_line')
+    if log_line:
+      event_data.log_line = log_line.lstrip().rstrip()
+
+    date_time_structure = self._GetValueFromStructure(structure, 'date_time')
+    date_time_parsed = self._ConvertTimeString(date_time_structure)
```

Comment:
since you already have the time elements parsed out in date_time_structure, I'd see if you could just pass them directly to the time_elements_tuple parameter.  see https://dfdatetime.readthedocs.io/en/latest/sources/api/dfdatetime.html#dfdatetime.time_elements.TimeElementsInMicroseconds

Reply:
Unfortunately the time zone formatting is non-standard and requires extra parsing that the dateutil library gives us for free

Reply:
Unfortunately the time zone format string is non standard and we need to use the dateutil parser to figure it out for us, which gives us back a datetime